### PR TITLE
OTA-1029: Add ClusterVersionOperator API

### PR DESCRIPTION
**Pull request by DavidHurta**
This pull request will introduce the `ClusterVersionOperator` API in the `DevPreviewNoUpgrade`  feature set as described in the https://github.com/openshift/enhancements/pull/1492 enhancement.

### CODE REVIEW

File: operator/v1alpha1/types_clusterversion.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	// +kubebuilder:default=Normal
+	OperatorLogLevel operatorv1.LogLevel `json:"operatorLogLevel,omitempty"`
+}
+
+// ClusterVersionOperatorStatus defines the observed status of the Cluster Version Operator.
+type ClusterVersionOperatorStatus struct {
+	// observedGeneration represents the most recent generation observed by the operator and specifies the version of
+	// the spec field currently being synced.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```
Comment (DavidHurta) [inline operator/v1alpha1/types_clusterversion.go:56]: The wording is updated based on the https://github.com/openshift/enhancements/pull/1492#discussion_r1835330712. The updated wording itself was not approved in the discussion and is for review. I am highlighting this as the change differs from the approved enhancement.

File: operator/v1alpha1/types_clusterversion.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+type ClusterVersionOperatorList struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard list's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ListMeta `json:"metadata"`
+
+	// items is a list of ClusterVersionOperators.
+	// +required
+	Items []ClusterVersionOperator `json:"items"`
```
Comment (DavidHurta) [inline operator/v1alpha1/types_clusterversion.go:74]: The `items` field was missing the `+required` or the `+optional` marker.

File: operator/v1alpha1/types_clusterversion.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+// +openshift:compatibility-gen:level=4
+type ClusterVersionOperatorList struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard list's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ListMeta `json:"metadata"`
+
+	// items is a list of ClusterVersionOperators.
+	// +required
```
Comment (JoelSpeed) [inline operator/v1alpha1/types_clusterversion.go:73]: Make this optional and add `omitempty` to the json tag please

File: operator/v1alpha1/types_clusterversion.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	// +kubebuilder:default=Normal
+	OperatorLogLevel operatorv1.LogLevel `json:"operatorLogLevel,omitempty"`
+}
+
+// ClusterVersionOperatorStatus defines the observed status of the Cluster Version Operator.
+type ClusterVersionOperatorStatus struct {
+	// observedGeneration represents the most recent generation observed by the operator and specifies the version of
+	// the spec field currently being synced.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```
Comment (JoelSpeed) [inline operator/v1alpha1/types_clusterversion.go:56]: We can add a validation to this to make sure that this only increases
```suggestion
	// +kubebuilder:validation:XValidation:rule="self >= oldSelf",message="observedGeneration should only increase"
	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```

File: operator/v1alpha1/types_clusterversion.go (hunk 3 above)
Comment (DavidHurta) [reply operator/v1alpha1/types_clusterversion.go]: Thanks, done in https://github.com/openshift/api/pull/2044/commits/92cd0ec3f270a6326de6ce64d4b127191754fa51.

File: operator/v1alpha1/types_clusterversion.go (hunk 4 above)
Comment (DavidHurta) [reply operator/v1alpha1/types_clusterversion.go:57]: Good point! The validation is now added in https://github.com/openshift/api/pull/2044/commits/d5e9f86abaad0de93340208e4e1cdb047d0bdb29. Additional API validation tests were added as well to test the validation.

**Comment by JoelSpeed** [comment]
/remove-label tide/merge-method-squash

You aren't supposed to use squash on payload contributing repositories like this. Please rebase and squash manually into atomic commits

**Pull request merged**